C++ bindings compile error

## Required information

**Operating system:**

Ubuntu 20.04.6 LTS
Linux ecm-unknown 4.9.337-l4t-r32.7.6+g44ff95afac8e #1 SMP PREEMPT Fri Apr 25 11:09:12 UTC 2025 aarch64 aarch64 aarch64 GNU/Linux

**Rust version:**
rustc 1.87.0 (17067e9ac 2025-05-09)

**Cargo version:**
cargo 1.87.0 (99624be96 2025-05-06)

**iceoryx2 version:**
main

**Detailed log output:**
```
In file included from /home/3dvl/workspace/libs/iceoryx2/iceoryx2-build/_deps/iceoryx-src/iceoryx_hoofs/vocabulary/include/iox/string.hpp:20,
                 from /home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/include/iox2/attribute.hpp:16,
                 from /home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/src/attribute.cpp:13:
/home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/include/iox2/enum_translation.hpp: In function 'constexpr typename iox::detail::extract_into_type<DestinationType>::type_t iox::from(SourceType) [with SourceType = int; DestinationType = iox2::SemanticStringError; typename iox::detail::extract_into_type<DestinationType>::type_t = iox2::SemanticStringError]':
/home/3dvl/workspace/libs/iceoryx2/iceoryx2-build/_deps/iceoryx-src/iceoryx_hoofs/reporting/include/iox/assertions.hpp:83:26: error: call to non-'constexpr' function 'void iox::er::forwardPanic(const iox::er::SourceLocation&, Message&&) [with Message = const char (&)[50]]'
     iox::er::forwardPanic(IOX_CURRENT_SOURCE_LOCATION, "Reached code that was supposed to be unreachable.")
     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/include/iox2/enum_translation.hpp:56:5: note: in expansion of macro 'IOX_UNREACHABLE'
     IOX_UNREACHABLE();
     ^~~~~~~~~~~~~~~
/home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/include/iox2/enum_translation.hpp: In function 'constexpr typename iox::detail::extract_into_type<DestinationType>::type_t iox::from(SourceType) [with SourceType = iox2::SemanticStringError; DestinationType = iox2_semantic_string_error_e; typename iox::detail::extract_into_type<DestinationType>::type_t = iox2_semantic_string_error_e]':
/home/3dvl/workspace/libs/iceoryx2/iceoryx2-build/_deps/iceoryx-src/iceoryx_hoofs/reporting/include/iox/assertions.hpp:83:26: error: call to non-'constexpr' function 'void iox::er::forwardPanic(const iox::er::SourceLocation&, Message&&) [with Message = const char (&)[50]]'
     iox::er::forwardPanic(IOX_CURRENT_SOURCE_LOCATION, "Reached code that was supposed to be unreachable.")
     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/include/iox2/enum_translation.hpp:70:5: note: in expansion of macro 'IOX_UNREACHABLE'
     IOX_UNREACHABLE();
     ^~~~~~~~~~~~~~~
/home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/include/iox2/enum_translation.hpp: In function 'constexpr typename iox::detail::extract_into_type<DestinationType>::type_t iox::from(SourceType) [with SourceType = int; DestinationType = iox2::ServiceType; typename iox::detail::extract_into_type<DestinationType>::type_t = iox2::ServiceType]':
/home/3dvl/workspace/libs/iceoryx2/iceoryx2-build/_deps/iceoryx-src/iceoryx_hoofs/reporting/include/iox/assertions.hpp:83:26: error: call to non-'constexpr' function 'void iox::er::forwardPanic(const iox::er::SourceLocation&, Message&&) [with Message = const char (&)[50]]'
     iox::er::forwardPanic(IOX_CURRENT_SOURCE_LOCATION, "Reached code that was supposed to be unreachable.")
     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/include/iox2/enum_translation.hpp:89:5: note: in expansion of macro 'IOX_UNREACHABLE'
     IOX_UNREACHABLE();
     ^~~~~~~~~~~~~~~
/home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/include/iox2/enum_translation.hpp: In function 'constexpr typename iox::detail::extract_into_type<DestinationType>::type_t iox::from(SourceType) [with SourceType = iox2::ServiceType; DestinationType = iox2_service_type_e; typename iox::detail::extract_into_type<DestinationType>::type_t = iox2_service_type_e]':
/home/3dvl/workspace/libs/iceoryx2/iceoryx2-build/_deps/iceoryx-src/iceoryx_hoofs/reporting/include/iox/assertions.hpp:83:26: error: call to non-'constexpr' function 'void iox::er::forwardPanic(const iox::er::SourceLocation&, Message&&) [with Message = const char (&)[50]]'
     iox::er::forwardPanic(IOX_CURRENT_SOURCE_LOCATION, "Reached code that was supposed to be unreachable.")
     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/include/iox2/enum_translation.hpp:102:5: note: in expansion of macro 'IOX_UNREACHABLE'
     IOX_UNREACHABLE();
     ^~~~~~~~~~~~~~~
/home/3dvl/workspace/libs/iceoryx2/iceoryx2/iceoryx2-ffi/cxx/include/iox2/enum_translation.hpp: In function 'constexpr typename iox::detail::extract_into_type<DestinationType>::type_t iox::from(SourceType) [with SourceType = int; DestinationType = iox2::NodeCreationFailure; typename iox::detail::extract_into_type<DestinationType>::type_t = iox2::NodeCreationFailure]':
/home/3dvl/workspace/libs/iceoryx2/iceoryx2-build/_deps/iceoryx-src/iceoryx_hoofs/reporting/include/iox/assertions.hpp:83:26: error: call to non-'constexpr' function 'void iox::er::forwardPanic(const iox::er::SourceLocation&, Message&&) [with Message = const char (&)[50]]'
     iox::er::forwardPanic(IOX_CURRENT_SOURCE_LOCATION, "Reached code that was supposed to be unreachable.")
     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
--> MORE ERRORS HERE <---
...
make[2]: *** [iceoryx2-ffi/cxx/CMakeFiles/iceoryx2-cxx-object-lib.dir/build.make:76: iceoryx2-ffi/cxx/CMakeFiles/iceoryx2-cxx-object-lib.dir/src/attribute.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1070: iceoryx2-ffi/cxx/CMakeFiles/iceoryx2-cxx-object-lib.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

**Observed result or behaviour:**
Compile error because IOX_UNREACHABLE() is not `constexpr`, but calling function is `constexpr`.

**Expected result or behaviour:**
C++ binding compilation should work.

**Conditions where it occurred / Performed steps:**
I followed instructions for building the C++ bindings.

@elBoberido Ah... compiler version was wrong. We pinned gcc version to 8.4.0 globally, because we use CUDA 10.2 on Nvidia Jetson and that is only compatible with gcc<=8.x. So the issue here is related to gcc 8 not 9. Sorry for that.

With gcc 9, I got another error:
```
[  0%] Built target iceoryx2-build-step
[  0%] Building C object examples/c/discovery/CMakeFiles/example_c_discovery.dir/src/main.c.o
[  1%] Linking C executable example_c_discovery
[  1%] Built target example_c_discovery
[  1%] Building C object examples/c/domains/CMakeFiles/example_c_domains_publisher.dir/src/publisher.c.o
[  2%] Linking C executable example_c_domains_publisher
/usr/bin/ld: ../../../rust/native/release/libiceoryx2_ffi.a(iceoryx2_pal_posix-215f632bc2e11e5b.iceoryx2_pal_posix.e889f182ce09a18b-cgu.1.rcgu.o): in function `iceoryx2_pal_posix::platform::mman::shm_open':
iceoryx2_pal_posix.e889f182ce09a18b-cgu.1:(.text._ZN18iceoryx2_pal_posix8platform4mman8shm_open17h33bd44627af4e123E+0x0): undefined reference to `shm_open'
/usr/bin/ld: ../../../rust/native/release/libiceoryx2_ffi.a(iceoryx2_pal_posix-215f632bc2e11e5b.iceoryx2_pal_posix.e889f182ce09a18b-cgu.1.rcgu.o): in function `iceoryx2_pal_posix::platform::mman::shm_unlink':
iceoryx2_pal_posix.e889f182ce09a18b-cgu.1:(.text._ZN18iceoryx2_pal_posix8platform4mman10shm_unlink17h8a1b95e387e3fe9cE+0x0): undefined reference to `shm_unlink'
/usr/bin/ld: ../../../rust/native/release/libiceoryx2_ffi.a(std-70a67151a7e3ef69.std.121563b87310a60c-cgu.0.rcgu.o): in function `std::sys::pal::unix::weak::fetch':
/rustc/17067e9ac6d7ecb70e50f92c1944e545188d2359/library/std/src/sys/pal/unix/weak.rs:141: undefined reference to `dlsym'
collect2: error: ld returned 1 exit status
make[2]: *** [examples/c/domains/CMakeFiles/example_c_domains_publisher.dir/build.make:98: examples/c/domains/example_c_domains_publisher] Error 1
make[1]: *** [CMakeFiles/Makefile2:678: examples/c/domains/CMakeFiles/example_c_domains_publisher.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

I could solve this by adding `$<$<PLATFORM_ID:Linux>:dl rt>` to the `target_link_libraries` of the `static-lib` target in `iceoryx2-ffi/c/CMakeLists.txt`. After that, compilation was working with gcc 9.4.

--- a/WORKSPACE.bazel
+++ b/WORKSPACE.bazel
@@ -141,12 +141,12 @@ bazel_skylib_workspace()
 
 
 # Load iceoryx rules
-ICEORYX_VERSION = "2.95.5"
+ICEORYX_VERSION = "2.95.6"
 
 maybe(
     name = "iceoryx",
     repo_rule = http_archive,
-    sha256 = "b68d0603ca39a852db5d12d8e83b0c7b560acdbe8d4cbcdfca1a07ada433bdd5",
+    sha256 = "745927c5e5e034cd840c1a18dff217818c042507b814d7bf001bcc61bcf5281b",
     strip_prefix = "iceoryx-{}".format(ICEORYX_VERSION),
     url = "https://github.com/eclipse-iceoryx/iceoryx/archive/v{}.tar.gz".format(ICEORYX_VERSION),
 )

--- a/doc/bazel/README.md
+++ b/doc/bazel/README.md
@@ -22,7 +22,7 @@ http_archive(
 
 
 # Load iceoryx rules
-ICEORYX_VERSION = "2.95.5"
+ICEORYX_VERSION = "2.95.6"
 
 maybe(
     name = "iceoryx",

--- a/doc/release-notes/iceoryx2-unreleased.md
+++ b/doc/release-notes/iceoryx2-unreleased.md
@@ -63,6 +63,8 @@
     [#875](https://github.com/eclipse-iceoryx/iceoryx2/issues/875)
 * Make `update_connections` public for all ports
     [#923](https://github.com/eclipse-iceoryx/iceoryx2/issues/923)
+* Fix linker error on linux
+    [#855](https://github.com/eclipse-iceoryx/iceoryx2/issues/855)
 
 ### Refactoring
 

--- a/iceoryx2-ffi/c/CMakeLists.txt
+++ b/iceoryx2-ffi/c/CMakeLists.txt
@@ -147,6 +147,7 @@ target_link_libraries(static-lib INTERFACE
     $<$<AND:$<PLATFORM_ID:Windows>,$<OR:$<C_COMPILER_ID:MSVC>,$<CXX_COMPILER_ID:MSVC>>>:bcrypt synchronization>
     $<$<PLATFORM_ID:Darwin>:stdc++>
     $<$<PLATFORM_ID:FreeBSD>:rt util>
+    $<$<PLATFORM_ID:Linux>:dl rt>
 )
 target_link_libraries(shared-lib INTERFACE
     iceoryx2-c::includes-only

--- a/iceoryx2-ffi/cxx/CMakeLists.txt
+++ b/iceoryx2-ffi/cxx/CMakeLists.txt
@@ -12,7 +12,7 @@
 
 cmake_minimum_required(VERSION 3.22)
 
-set(ICEORYX_HOOFS_VERSION 2.95.5)
+set(ICEORYX_HOOFS_VERSION 2.95.6)
 include(${CMAKE_CURRENT_LIST_DIR}/cmake/fetch-iceoryx-hoofs.cmake)
 
 project(iceoryx2-cxx VERSION ${IOX2_VERSION_STRING} LANGUAGES CXX)

--- a/iceoryx2-ffi/cxx/README.md
+++ b/iceoryx2-ffi/cxx/README.md
@@ -38,7 +38,7 @@ specify the path to the install directory with `-DCMAKE_PREFIX_PATH`.
 `iceoryx_hoofs` can be build with this steps:
 
 ```bash
-git clone --depth 1 --branch v2.95.5 https://github.com/eclipse-iceoryx/iceoryx.git target/iceoryx/src
+git clone --depth 1 --branch v2.95.6 https://github.com/eclipse-iceoryx/iceoryx.git target/iceoryx/src
 
 cmake -S target/iceoryx/src/iceoryx_platform -B target/iceoryx/build/platform -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=target/iceoryx/install
 cmake --build target/iceoryx/build/platform

--- a/internal/scripts/ci_build_and_install_iceoryx_hoofs.ps1
+++ b/internal/scripts/ci_build_and_install_iceoryx_hoofs.ps1
@@ -21,7 +21,7 @@ $ErrorActionPreference = "Stop"
 
 $NUM_JOBS = (Get-WmiObject Win32_processor).NumberOfLogicalProcessors
 
-git clone --depth 1 --branch v2.95.5 https://github.com/eclipse-iceoryx/iceoryx.git target/iceoryx/src
+git clone --depth 1 --branch v2.95.6 https://github.com/eclipse-iceoryx/iceoryx.git target/iceoryx/src
 
 switch ($mode) {
     "release" {

--- a/internal/scripts/ci_build_and_install_iceoryx_hoofs.sh
+++ b/internal/scripts/ci_build_and_install_iceoryx_hoofs.sh
@@ -26,7 +26,7 @@ elif [[ "$OSTYPE" == "darwin"* ]]; then
     NUM_JOBS=$(sysctl -n hw.ncpu)
 fi
 
-git clone --depth 1 --branch v2.95.5 https://github.com/eclipse-iceoryx/iceoryx.git target/iceoryx/src
+git clone --depth 1 --branch v2.95.6 https://github.com/eclipse-iceoryx/iceoryx.git target/iceoryx/src
 
 cmake -S target/iceoryx/src/iceoryx_platform -B target/iceoryx/build/platform -DBUILD_SHARED_LIBS=OFF -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=target/iceoryx/install
 cmake --build target/iceoryx/build/platform -j$NUM_JOBS

--- a/package.xml
+++ b/package.xml
@@ -9,7 +9,7 @@
     <url type="bugtracker">https://github.com/eclipse-iceoryx/iceoryx2/issues</url>
     <url type="repository">https://github.com/eclipse-iceoryx/iceoryx2</url>
     <buildtool_depend>cmake</buildtool_depend>
-    <depend version_gte="2.95.5">iceoryx_hoofs</depend>
+    <depend version_gte="2.95.6">iceoryx_hoofs</depend>
     <export>
         <build_type>cmake</build_type>
     </export>

